v-next/hardhat-errors/src/errors.ts

```diff
@@ -130,68 +186,27 @@ function getErrorCode(errorDescriptor: ErrorDescriptor): string {
  */
 export function applyErrorMessageTemplate(
   template: string,
-  values: ErrorMessageTemplateArguments,
+  values: Record<string, ErrorMessageTemplateValue>,
 ): string {
-  return _applyErrorMessageTemplate(template, values, false);
-}
-
-function _applyErrorMessageTemplate(
-  template: string,
-  values: ErrorMessageTemplateArguments,
-  isRecursiveCall: boolean,
-): string {
-  if (!isRecursiveCall) {
-    for (const variableName of Object.keys(values)) {
-      assertHardhatInvariant(
-        /^[a-zA-Z][a-zA-Z0-9]*$/.test(variableName),
-        `Trying to apply error template but variable "${variableName}" is invalid. Variable names can only include ascii letters and numbers, and start with a letter.`,
-      );
-
-      const variableTag = `%${variableName}%`;
-
-      assertHardhatInvariant(
-        template.includes(variableTag),
-        `Trying to apply error template but variable "${variableName}" is not present in the template.`,
-      );
-    }
-  }
-
-  if (template.includes("%%")) {
-    return template
-      .split("%%")
-      .map((part) => _applyErrorMessageTemplate(part, values, true))
-      .join("%");
-  }
-
-  for (const variableName of Object.keys(values)) {
-    let value: string;
-
+  return template.replaceAll(/{(.*?)}/g, (_match, variableName) => {
```

Comment:
should this match the empty group `{}`?

Reply:
Good catch! But after giving it a few minutes, I think it's better if it matches the behavior of the types. To avoid matching the empty group, the types will be significantly more complex. 

I'll add a test for this case.

Reply:
Added the tests in [409dc51](https://github.com/NomicFoundation/hardhat/pull/5283/commits/409dc510353906bf50140abbec1319723fd7be8c)